packages/dashboard-frontend/src/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/index.tsx

```diff
@@ -229,52 +231,81 @@ class StepApplyDevfile extends AbstractLoaderStep<Props, State> {
       return true;
     }
 
-    if (!shouldCreate || this.state.lastError instanceof Error) {
-      if (this.state.lastError instanceof Error) {
-        throw this.state.lastError;
-      }
+    if (shouldCreate === false) {
       throw new Error('The workspace creation unexpectedly failed.');
     }
 
-    if (devfile === undefined) {
-      if (factoryResolver === undefined) {
-        const _devfile = defaultDevfile;
-        if (_devfile === undefined) {
+    // factory resolving failed in the previous step
+    // hence we have to proceed with the default devfile
+    if (factoryResolver === undefined) {
+      if (devfile === undefined) {
+        if (defaultDevfile === undefined) {
           throw new Error('Failed to resolve the default devfile.');
         }
-        this.updateCurrentDevfile(_devfile);
+        this.updateCurrentDevfile(defaultDevfile);
+      } else {
         try {
-          await this.createWorkspaceFromDevfile(_devfile);
+          await this.createWorkspaceFromDevfile(devfile);
         } catch (e) {
           const errorMessage = common.helpers.errors.getMessage(e);
           throw new CreateWorkspaceError(errorMessage);
         }
+      }
+      return false;
+    }
+
+    // the user devfile is invalid and caused creation error
+    // so we have to proceed with the default devfile
+    if (continueWithDefaultDevfile === true) {
+      if (defaultDevfile === undefined) {
+        throw new Error('Failed to resolve the default devfile.');
+      }
+
+      if (devfile === undefined) {
+        const _devfile = cloneDeep(defaultDevfile);
```

Comment:
The ```cloneDeep``` selector will return a new instance each time:

```
        const devfile = load(devfileContent) as devfileApi.Devfile;
        // propagate default components
        if (!devfile.components || devfile.components.length === 0) {
          devfile.components = defaultComponents;
        }
        return devfile;
```

https://github.com/eclipse-che/che-dashboard/blob/loader-steps-conditions/packages/dashboard-frontend/src/store/DevfileRegistries/selectors.ts#L81

Why do we need this ```cloneDeep(defaultDevfile);```?

Reply:
I think it makes sense to get rid of `cloneDeep`, thanks.